test/end2end_test.go

```diff
@@ -145,6 +145,8 @@ type testServer struct {
 	setHeaderOnly      bool   // whether to only call setHeader, not sendHeader.
 	multipleSetTrailer bool   // whether to call setTrailer multiple times.
 	unaryCallSleepTime time.Duration
+	earlyNil           bool // whether to return nil without calling SendAndClose().
+	recvAfterClose     bool // whether to call Recv() after calling SendAndClose().
```

Comment:
We really really don't want to extend `testServer`.  Could you use the `stubserver` package instead?  `testServer` has a lot of bad properties (especially: it splits the testing code into two different places: the test and the `testServer`), and this only complicates it further by adding more ways to conditionalize its behavior.

You can follow this kind of example to use `stubserver` instead:

https://github.com/grpc/grpc-go/blob/d48317fafe7361f172fc243bbde7e5594f12ebd5/test/end2end_test.go#L4485

Also please add a comment on `testServer` that it should not be used for any new tests so that this doesn't happen again.

Sorry!